osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/DlmsPushNotificationDecoder.java

```diff
@@ -62,22 +85,184 @@ private void reset() {
+         *  00                                                       // Date-time (empty)
+         *  02 02                                                    // Structure with 2 elements
+         *  09 11 45 58 58 58 58 31 32 33 34 35 36 37 38 39 30 31 32 // Equipment id EXXXX123456789012
+         *  09 06 00 00 19 09 00 FF                                  // Logical name: Push setup schedule
+         *
+         *  0F                                                       // Data-notification
+         *  00 00 00 01                                              // Long-invoke-id-and-priority (can be ignored)
+         *  00                                                       // Date-time (empty)
+         *  02 03                                                    // Structure with 3 elements
+         *  09 11 45 58 58 58 58 31 32 33 34 35 36 37 38 39 30 31 32 // Equipment id EXXXX123456789012
+         *  09 06 00 01 19 09 00 FF                                  // Logical name: Push setup alarms
+         *  06 00 00 00 02                                           // Alarm register, with Replace battery set
+         *
+         *  Notes:
+         *  - For SMR5 alarms, we get 8 additional addressing bytes in front of the alarm. These bytes can be
+         *  ignored here.
+         *  - To check if the alarm is in DSMR4 or SMR5 format, check the 9th byte (at index 8). If it is 0F, then it
+         *  is SMR5, otherwise it is DSMR4, because the 9th byte in DSMR4 is in the identifier and this should be a
+         *  number or a character, so it can't be ASCII code 0F.
+         * */
+
+        DlmsPushNotification pushNotification;
 
-        LOGGER.info("Decoding state: {}", state);
+        // Determine whether the alarm is in DSMR4 or SMR5 format.
+        boolean smr5alarm = buffer.getByte(8) == 0x0F;
 
-        switch (state) {
-        case EQUIPMENT_IDENTIFIER:
-            this.decodeEquipmentIdentifier(buffer);
-            return this.setCheckpointAndContinueDecode(ctx, channel, buffer, DecodingState.DATA_OBJECT);
-        case DATA_OBJECT:
-            this.decodeReceivedData(buffer);
-            return this.buildPushNotification();
-        default:
-            throw new UnknownDecodingStateException(String.valueOf(state));
+        LOGGER.info("Decoding state: {}, SMR5 alarm: {}", state, smr5alarm);
+
+        if (smr5alarm) {
+            pushNotification = decodeSmr5alarm(buffer);
+        } else {
+            switch (state) {
+            case EQUIPMENT_IDENTIFIER:
+                this.decodeEquipmentIdentifier(buffer);
+                return this.setCheckpointAndContinueDecode(ctx, channel, buffer, DecodingState.DATA_OBJECT);
+            case DATA_OBJECT:
+                this.decodeReceivedData(buffer);
+                pushNotification = this.buildPushNotification();
+                break;
+            default:
+                throw new UnknownDecodingStateException(String.valueOf(state));
+            }
```

Comment:
This class handles different decoding cases using a switch statement (as well as it being 416 lines of code). This screams: "Refactor me!" I suggest splitting it in separate classes, at least for SMR5 and DSMR4 but maybe also for the EQUIPMENT_IDENTIFIER and DATA_OBJECT state logic.

Reply:
Refactored into 2 seperate classes. Removed decoding state in DSMR4 alarm decoding.